internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,497 @@
+
+	this.WorkErrors = workErrors
+}
+
+func (this JobOutput[InputType, OutputType]) SetResultItems(resultItems []OutputType) {
+	lockmanager.Lock(this.JobID)
+	defer lockmanager.Unlock(this.JobID)
+
+	lockKey := fmt.Sprintf(JOB_OUTPUT_LOCK_FORMAT, this.JobID, JOB_OUTPUT_RESULT_ITEMS)
+	lockmanager.Lock(lockKey)
+	defer lockmanager.Unlock(lockKey)
+
+	this.ResultItems = resultItems
+}
+
+func (this JobOutput[InputType, OutputType]) SetRemainingItems(remainingItems []InputType) {
+	lockmanager.Lock(this.JobID)
+	defer lockmanager.Unlock(this.JobID)
+
+	lockKey := fmt.Sprintf(JOB_OUTPUT_LOCK_FORMAT, this.JobID, JOB_OUTPUT_REMAINING_ITEMS)
+	lockmanager.Lock(lockKey)
+	defer lockmanager.Unlock(lockKey)
+
+	this.RemainingItems = remainingItems
+}
+
+func (this JobOutput[InputType, OutputType]) SetRunTime(runTime int64) {
+	lockmanager.Lock(this.JobID)
+	defer lockmanager.Unlock(this.JobID)
+
+	lockKey := fmt.Sprintf(JOB_OUTPUT_LOCK_FORMAT, this.JobID, JOB_OUTPUT_RUN_TIME)
+	lockmanager.Lock(lockKey)
+	defer lockmanager.Unlock(lockKey)
+
+	this.RunTime = runTime
+}
+
+// Given a customized Job, Job.Run() processes input items in a parallel pool of workers.
+// Returns the collected results in a JobOutput.
+// If the context is canceled during execution, returns nil.
+// When not waiting for completion, the JobOutput may still be modified until the Done channel is closed.
+// Returning a copy of results gives immediate access to stable results but the final results will not be accessible later.
+func (this *Job[InputType, OutputType]) Run() *JobOutput[InputType, OutputType] {
+	done := make(chan any)
+
+	output := JobOutput[InputType, OutputType]{
+		Done:           done,
+		ResultItems:    make([]OutputType, 0, len(this.WorkItems)),
+		RemainingItems: this.WorkItems,
+		RunTime:        0,
+		WorkErrors:     make(map[int]error, 0),
+		JobID:          util.UUID(),
+	}
+
+	err := this.Validate()
+	if err != nil {
+		close(done)
+
+		output.Error = fmt.Errorf("Failed to validate job: '%v'.", err)
+		return &output
+	}
+
+	// If we are overwriting records, remove all the old records.
+	if this.OverwriteRecords && !this.DryRun && this.RemoveStorageFunc != nil {
+		output.Error = this.RemoveStorageFunc(this.WorkItems)
+		if output.Error != nil {
+			close(done)
+
+			return &output
+		}
+	}
+
+	// Query for stored records.
+	if !this.OverwriteRecords && this.RetrieveFunc != nil {
+		output.ResultItems, output.RemainingItems, output.Error = this.RetrieveFunc(this.WorkItems)
+		if output.Error != nil {
+			close(done)
+
+			return &output
+		}
+	}
+
+	if this.WaitForCompletion {
+		this.run(&output, true)
+
+		close(done)
+
+		// If the context was canceled during execution, return immediately.
+		if this.Context.Err() != nil {
+			return nil
+		}
+	} else if this.ReturnCopy {
+		go func() {
+			backgroundOutput := &JobOutput[InputType, OutputType]{
+				RemainingItems: output.GetRemainingItems(),
+			}
+
+			this.run(backgroundOutput, false)
+
```

Comment:
Remove this newline.
Careful not to separate a function call from its error check.